pkg/reconciler/revision/resources/deploy.go

```diff
@@ -156,45 +156,69 @@ func makePodSpec(rev *v1.Revision, loggingConfig *logging.Config, tracingConfig
 	return podSpec, nil
 }
 
-// BuildUserContainer makes a container from the Revision template.
-func BuildUserContainer(rev *v1.Revision) *corev1.Container {
-	userContainer := rev.Spec.GetContainer().DeepCopy()
-	// Adding or removing an overwritten corev1.Container field here? Don't forget to
-	// update the fieldmasks / validations in pkg/apis/serving
+// BuildUserContainers makes a container from the Revision template.
+func BuildUserContainers(rev *v1.Revision) []corev1.Container {
+	containers := make([]corev1.Container, 0, len(rev.Spec.PodSpec.Containers))
 
-	userContainer.VolumeMounts = append(userContainer.VolumeMounts, varLogVolumeMount)
-	userContainer.Lifecycle = userLifecycle
-	userPort := getUserPort(rev)
-	userPortInt := int(userPort)
-	userPortStr := strconv.Itoa(userPortInt)
-	// Replacement is safe as only up to a single port is allowed on the Revision
-	userContainer.Ports = buildContainerPorts(userPort)
-	userContainer.Env = append(userContainer.Env, buildUserPortEnv(userPortStr))
-	userContainer.Env = append(userContainer.Env, getKnativeEnvVar(rev)...)
-	// Explicitly disable stdin and tty allocation
-	userContainer.Stdin = false
-	userContainer.TTY = false
+	var container corev1.Container
+	for i := range rev.Spec.PodSpec.Containers {
```

Comment:
nit: since the var is only used inside the loop might be a bit nicer to declare it there

```suggestion
	for i := range rev.Spec.PodSpec.Containers {
		var container corev1.Container
```